File: common/src/main/java/dev/mayaqq/estrogen/networking/messages/c2s/SetChestConfigPacket.java
package dev.mayaqq.estrogen.networking.messages.c2s;

import com.teamresourceful.resourcefullib.common.network.Packet;
import com.teamresourceful.resourcefullib.common.network.base.PacketType;
import com.teamresourceful.resourcefullib.common.network.base.ServerboundPacketType;
import com.teamresourceful.resourcefullib.common.network.defaults.CodecPacketType;
import dev.mayaqq.estrogen.Estrogen;
import dev.mayaqq.estrogen.config.ChestConfig;
import dev.mayaqq.estrogen.config.PlayerEntityExtension;
import dev.mayaqq.estrogen.networking.EstrogenNetworkManager;
import dev.mayaqq.estrogen.networking.messages.s2c.ChestConfigPacket;
import net.minecraft.world.entity.player.Player;

import java.util.function.Consumer;

public record SetChestConfigPacket(ChestConfig config) implements Packet<SetChestConfigPacket> {

    public static ServerboundPacketType<SetChestConfigPacket> TYPE = new Type();

    @Override
    public PacketType<SetChestConfigPacket> type() {
        return TYPE;
    }

    private static class Type extends CodecPacketType.Server<SetChestConfigPacket> {

        public Type() {
            super(
                    SetChestConfigPacket.class,
                    Estrogen.id("chest_feature_config"),
                    ChestConfig.BYTE_CODEC.map(SetChestConfigPacket::new, SetChestConfigPacket::config)
            );
        }

        @Override
        public Consumer<Player> handle(SetChestConfigPacket message) {
            return (player) -> {
                ((PlayerEntityExtension) player).estrogen$setChestConfig(message.config);
                EstrogenNetworkManager.CHANNEL.sendToPlayersInLevel(new ChestConfigPacket(player.getUUID(), message.config), player.level());
            };
        }
    }
}

File: src/common/main/kotlin/dev/mayaqq/estrogen/network/messages/c2s/SetChestConfigPacket.kt
package dev.mayaqq.estrogen.network.messages.c2s

import dev.mayaqq.cynosure.entities.PlayerLookup
import dev.mayaqq.cynosure.network.Packet
import dev.mayaqq.cynosure.network.SerializablePacket
import dev.mayaqq.cynosure.network.ServerNetworkContext
import dev.mayaqq.estrogen.config.types.ChestConfig
import dev.mayaqq.estrogen.content.EstrogenEffects
import dev.mayaqq.estrogen.content.effects.EstrogenEffect
import dev.mayaqq.estrogen.injection.chestConfig
import dev.mayaqq.estrogen.network.EstrogenNetwork
import dev.mayaqq.estrogen.network.messages.s2c.ChestConfigPacket

@SerializablePacket("set_chest_config")
data class SetChestConfigPacket(val config: ChestConfig) : Packet.Serverbound {
    override fun ServerNetworkContext.handle() = execute {
        sender.chestConfig = config

        for (player in PlayerLookup.tracking(sender)) {
            EstrogenEffect.sendPlayerStatusEffect(player, EstrogenEffects.Estrogen, sender)
            player.chestConfig?.let {
                EstrogenNetwork.sendToPlayer(ChestConfigPacket(player.uuid, it), sender)
            }
        }
    }
}